Add aspect ratio locking and flip support to skin editor

peppy opened the pull request:
- [x] Depends on https://github.com/ppy/osu/pull/12762

This should be considered temporary implementation along with the rest, intended to get the initial release into a roughly usable state. Please be easy on the review as it will get replaced with an abstracted version of `OsuSelectionHandler` in the very near future.

peppy committed:
Add aspect ratio locking and flip support to skin editor

peppy committed:
Fix display of skin blueprints when flipped

peppy committed:
Fix skin blueprint box drawing incorrectly when both scale and rotation are applied

peppy committed:
Merge branch 'master' into skin-blueprint-aspect-lock

bdach commented on osu.Game/Skinning/Editor/SkinSelectionHandler.cs:
```diff
@@ -139,6 +154,15 @@ private static void adjustScaleFromAnchor(ref Vector2 scale, Anchor reference)
+            // for now aspect lock scale adjustments that occur at corners.
+            if (!reference.HasFlagFast(Anchor.x1) && !reference.HasFlagFast(Anchor.y1))
+            {
+                if (reference.HasFlagFast(Anchor.x0) || reference.HasFlagFast(Anchor.x2))
+                    scale.Y = scale.X;
+                else
+                    scale.X = scale.Y;
```
What was the intention here? Asking because the `else` branch seems effectively dead here. I've breakpointed on it and it seems to be unreachable. To explain what I mean:

```
    x0  x1  x2
    |   |   |
y0 -O---O---O-
    |   |   |
y1 -O---+---O-
    |   |   |
y2 -O---O---O-
    |   |   |
```

The `O`s indicate where the reference anchors are on a selection box. The first conditional eliminates the middle ones, which makes sense. But after excluding them from further deliberations (marking via `X`):

```
    x0  x1  x2
    |   |   |
y0 -O---X---O-
    |   |   |
y1 -X---+---X-
    |   |   |
y2 -O---X---O-
    |   |   |
```

all the reference anchors left have `Anchor.x0` or `Anchor.x2`...

peppy replied:
i'm not actually sure, but it felt okay. this code will likely be removed/replaced within the next week, so just killing that line is fine for now.

bdach replied:
I *think* what may have been wanted is

```csharp
if (scale.X != 0 && scale.Y != 0)
    scale.X = scale.Y = Math.Max(scale.X, scale.Y);
else if (scale.X != 0)
    scale.Y = scale.X;
else if (scale.Y != 0)
    scale.X = scale.Y;
```

The misleading part about all this is that `scale` is all deltas, not the actual full final scale.

peppy replied:
I tried that initially but it doesn't work amazingly because as you say, it's arriving as deltas and therefore not in a stable state.

mThats why i chose to leave all the ￼decisions based on the dragged anchor alone.

bdach replied:
Fair enough. I'll just kill the `else` branch and leave a TODO for now, then...

bdach committed:
Remove dead branch and mark implementation as temporary

The previous implementation was checking if the `x0` or `x2` anchors
were selected to decide on which way to transfer the drawable's scale,
but that check actually ends up being always true for corner anchors. To
visualise, this is how the corner anchors correspond to `Anchor` flags:

    x0  x1  x2
    |   |   |
y0 -O---O---O-
    |   |   |
y1 -O---+---O-
    |   |   |
y2 -O---O---O-
    |   |   |

The Os indicate where the reference anchors are on a selection box.
The first conditional eliminates the middle ones, which makes sense.
But after excluding them from further deliberations (marking via X):

    x0  x1  x2
    |   |   |
y0 -O---X---O-
    |   |   |
y1 -X---+---X-
    |   |   |
y2 -O---X---O-
    |   |   |

The remaining anchors always have `x0` or `x2` set. So to avoid
confusion, just always transfer one way for now. At some point this
should be torn out in favour of an actual implementation of the desired
behaviour.

bdach committed:
Merge branch 'master' into skin-blueprint-aspect-lock
